libraries/AP_HAL_SITL/Semaphores.cpp

```diff
@@ -76,4 +76,151 @@ bool Semaphore::take_nonblocking()
     return false;
 }
 
+
+#if AP_COUNTSEM_NATIVE_SEM_T
+/*
+  counting semaphores using sem_t
+ */
+CountingSemaphore::CountingSemaphore(uint8_t initial_count) :
+    AP_HAL::CountingSemaphore(initial_count)
+{
+    sem_init(&_sem, 0, initial_count);
+}
+
+bool CountingSemaphore::wait(uint32_t timeout_us)
+{
+    if (hal.scheduler->in_main_thread() ||
+        Scheduler::from(hal.scheduler)->semaphore_wait_hack_required()) {
+        /*
+          when in the main thread we need to do a busy wait to ensure
+          the clock advances
+         */
+        uint64_t end_us = AP_HAL::micros64() + timeout_us;
+        do {
+            if (sem_trywait(&_sem) == 0) {
+                return true;
+            }
+            hal.scheduler->delay_microseconds(10);
+        } while (AP_HAL::micros64() < end_us);
+        return false;
+    }
+    
+    if (timeout_us == 0) {
+        return sem_trywait(&_sem) == 0;
+    }
+    struct timespec ts;
+    if (clock_gettime(CLOCK_REALTIME, &ts) != 0) {
+        return false;
+    }
+    ts.tv_sec += timeout_us/1000000UL;
+    ts.tv_nsec += (timeout_us % 1000000U) * 1000UL;
+    if (ts.tv_nsec >= 1000000000L) {
+        ts.tv_sec++;
+        ts.tv_nsec -= 1000000000L;
+    }
+    return sem_timedwait(&_sem, &ts) == 0;
+}
+
+bool CountingSemaphore::wait_blocking(void)
+{
+    return sem_wait(&_sem) == 0;
+}
+
+void CountingSemaphore::signal(void)
+{
+    sem_post(&_sem);
```

Comment:
same point as discussed for AP_HAL_ChibiOS::CountinSemaphore::signal. We should probably be limiting the value to 1. So we don't create multiple runs due to multiple signals between waits.

Reply:
changed to binary